src/nbl/video/utilities/IUtilities.cpp

```diff
@@ -669,15 +163,662 @@ void IUtilities::updateImageViaStagingBuffer(
+{
+    asset::TexelBlockInfo dstImageTexelBlockInfo(dstImageFormat);
+    assert(dstImageTexelBlockInfo.getBlockByteSize()>0u);
+    auto texelBlockDim = dstImageTexelBlockInfo.getDimension();
+    size_t memoryNeededForRemainingRegions = 0ull;
+    for (uint32_t i = currentRegion; i < regions.size(); ++i)
+    {
+        memoryNeededForRemainingRegions = core::alignUp(memoryNeededForRemainingRegions, bufferOffsetAlignment);
+
+        const asset::IImage::SBufferCopy & region = regions[i];
+
+        auto subresourceSize = dstImage->getMipSize(region.imageSubresource.mipLevel);
+
+        assert(static_cast<uint32_t>(region.imageSubresource.aspectMask) != 0u);
+        assert(core::isPoT(static_cast<uint32_t>(region.imageSubresource.aspectMask)) && "region.aspectMask should only have a single bit set.");
+        // Validate Region
+        // canTransferMipLevelsPartially
+        if(!canTransferMipLevelsPartially)
+        {
+            assert(region.imageOffset.x == 0 && region.imageOffset.y == 0 && region.imageOffset.z == 0);
+            assert(region.imageExtent.width == subresourceSize.x && region.imageExtent.height == subresourceSize.y && region.imageExtent.depth == subresourceSize.z);
+        }
+
+        // region.imageOffset.{xyz} should be multiple of minImageTransferGranularity.{xyz} scaled up by block size
+        bool isImageOffsetAlignmentValid =
+            (region.imageOffset.x % (minImageTransferGranularity.width  * texelBlockDim.x) == 0) &&
+            (region.imageOffset.y % (minImageTransferGranularity.height * texelBlockDim.y) == 0) &&
+            (region.imageOffset.z % (minImageTransferGranularity.depth  * texelBlockDim.z) == 0);
+        assert(isImageOffsetAlignmentValid);
+
+        // region.imageExtent.{xyz} should be multiple of minImageTransferGranularity.{xyz} scaled up by block size,
+        // OR ELSE (region.imageOffset.{x/y/z} + region.imageExtent.{width/height/depth}) MUST be equal to subresource{Width,Height,Depth}
+        bool isImageExtentAlignmentValid = 
+            (region.imageExtent.width  % (minImageTransferGranularity.width  * texelBlockDim.x) == 0 || (region.imageOffset.x + region.imageExtent.width   == subresourceSize.x)) && 
+            (region.imageExtent.height % (minImageTransferGranularity.height * texelBlockDim.y) == 0 || (region.imageOffset.y + region.imageExtent.height  == subresourceSize.y)) &&
+            (region.imageExtent.depth  % (minImageTransferGranularity.depth  * texelBlockDim.z) == 0 || (region.imageOffset.z + region.imageExtent.depth   == subresourceSize.z));
+        assert(isImageExtentAlignmentValid);
+
+        bool isImageExtentAndOffsetValid = 
+            (region.imageExtent.width + region.imageOffset.x <= subresourceSize.x) &&
+            (region.imageExtent.height + region.imageOffset.y <= subresourceSize.y) &&
+            (region.imageExtent.depth + region.imageOffset.z <= subresourceSize.z);
+        assert(isImageExtentAndOffsetValid);
+
```

Comment:
you shouldn't assert these, you shouldn't have been able to construct an iterator in the first place with invalid regions due to:
- whole mip requirement
- offset and extent not being multiples of granularity or texel block dimensions
etc.

there's this little nice thing
https://github.com/Devsh-Graphics-Programming/Nabla/blob/master/include/nbl/asset/IImage.h#L473 
mauybe you could add a `const VkExtent3D& transferGranularity={1,1,1}` as  extra argument

Reply:
the nice thing is that there is also one in IGPUImage that makes more sense to Add!
https://github.com/Devsh-Graphics-Programming/Nabla/blob/da32613583c96b0dbef2a2191dda7d61618090cc/include/nbl/video/IGPUImage.h#L61

Reply:
For future reference:
I added validaiton of minImageTransferGranularity and other checks to `IGPUImage::validateCopies` but I couldn't use this since the ICPUBuffer (srcBuffer) is not promoted and hence is not valid to be used in copy (the function doesn't know that the upload utility will handle this)
So i made another function used to validate "extent" and "offset" with `minImageTransferGranularity` in IGPUImage and use that directly to verify All regions